### PR TITLE
Fixes #466: Implements the enumerate generator methods

### LINKED ISSUE #466: Add Iter... methods to WBEMConnection to hide open... vs enumerate operations
We will want to extend the enumerates so that the user does not have to make decisions about whether to use pull or not.

If we do it at the WBEMConnection level, the first enumerate request could test for existence of pull operations and set a flag that would last for the life of the WBEMConnection.  I think that the it is logical to do this at the WBEM.

This would also require that the user be able to set some configuration information:
1. interoperation timeout
2. Max objects on each pull

These could be added to the WBEM connection optional variables with a default.

new new operations would probably be based on a generator so that the user would call the operation and get a generator returned.

### PR DESCRIPTION
**Pull request by KSchopmeyer**
**STATUS 11 Dec:** for review: All functions implemented along with a _use... parameter for each.  Removed the method to set use_pull_operations.  Some initial code for these operations is now in run_cim operations but have more work there.  I have not yet sorted out how to test this through mock yet. First goal is complete test with run_cimoperations.py I have at least a day of adding tests just in run_cim_operations.py

Added almost all basic tests (except invoke method) to run_cimoperations.py so we are exercising at least the main paths. Working now to try to get them into mock but still trying to sort out how to handle the generate response(probably use it to drive a for loop)

**Discussion:** The IterQueryInstances works but want to discuss the fact that what I did is a hack. Using generators has a number of limitations, specifically it is not easy to create results that return the generator type and some other entity.  I created a class that is returned and has two properties, the query_result_class (on None) and the generator object.

However, making the chain so that I can feed this return class from the pull responses proved difficutltso for now I get all and then return them IterQueryInstancesReturn. The alternative is another class wrapping the whole implementation of IterQueryInstances so that we could wrap multiple levels of generators and I ran out of time.

**DISCUSSION:** restructure with iter function in separate class: See last comment for my conclusion after implementing. **CONCLUSION:** Agreed not to do this

There is also a sample program in examples right now.  The example does execute but is incomplete.

Known Questions:
2. One of the inefficiencies of this is that for every connection we go through the pull dance first try pull and then if pull fails, non-pull. For scripts, where there is just one connection, this is not a real problem. However,  it does assume that the user makes a WBEMConnection with a WEBServer and keeps using it to be efficient. No solution. This is just a note for now.

### CODE REVIEW

File: pywbem/cim_operations.py [hunk 1]
```diff
@@ -1709,6 +1721,269 @@ def OpenEnumerateInstancePaths(self, ClassName, namespace=None,
...
+
+            Exceptions described in :class:`~pywbem.WBEMConnection`.
+
+        Returns:
+
+          A Python iterator object. Instances can be retrieved
+          by iterating through the iterator.
+
+        Example:
+            enum_gen = conn.EnumerateInstancesWithPathGen('CIM_Blah')
```
Comment (andy-maier) [inline pywbem/cim_operations.py]: - The function is now called differently.
- The return value is an iterator not a generator and the variable storing it should not end with `_gen`. Maybe just  `instances`?

File: pywbem/cim_operations.py [hunk 2]
```diff
@@ -1709,6 +1721,269 @@ def OpenEnumerateInstancePaths(self, ClassName, namespace=None,
...
+                if inst.get('thisPropertyName') == 0
+                    enum_get.close()
+                    break;
+                else:
+                    print('instance %s' % inst.tomof())
+
+        """
+
+        if MaxObjectCount is not None and MaxObjectCount <= 0:
+            raise ValueError('MaxObjectCount=%s. Invalid" le 0' % \
```
Comment (andy-maier) [inline pywbem/cim_operations.py]: There's probably something wrong in the format string.

File: pywbem/cim_operations.py [hunk 3]
```diff
@@ -1709,6 +1721,269 @@ def OpenEnumerateInstancePaths(self, ClassName, namespace=None,
...
+                        pull_result = self.PullInstancesWithPath(
+                            pull_result.context, MaxObjectCount=MaxObjectCount)
+
+                        while inst in pull_result.instances:
+                            yield inst
+                    return
+
+                # if Not supported, set flag and try alternative operation.
+                except CIMError as ce:
+                    if ce.args[0] != CIM_ERR_NOT_SUPPORTED:
```
Comment (andy-maier) [inline pywbem/cim_operations.py]: Use `ce.status_code` instead.

File: pywbem/cim_operations.py [hunk 4]
```diff
@@ -1709,6 +1721,269 @@ def OpenEnumerateInstancePaths(self, ClassName, namespace=None,
...
+                        self.has_pull_operations = False
+                    else:
+                        raise
```
Comment (andy-maier) [inline pywbem/cim_operations.py]: Here is a logic error. The description says we want to try pull enumerations if `has_pull_operations` is None, and set it to True if that succeeds. That second part does not happen in the current code, it can be done in an else path of the try/except.

File: pywbem/cim_operations.py [hunk 5]
```diff
@@ -1709,6 +1721,269 @@ def OpenEnumerateInstancePaths(self, ClassName, namespace=None,
...
+
```
Comment (andy-maier) [inline pywbem/cim_operations.py]: How about inserting an `assert self.has_pull_operations == False` to document this?

File: pywbem/cim_operations.py [hunk 6]
```diff
@@ -1709,6 +1721,269 @@ def OpenEnumerateInstancePaths(self, ClassName, namespace=None,
...
+            # Alternate request if Pull not implemented. This does not allow
+            # the FilterQuery
+            if FilterQuery is not None or FilterQueryLanguage is not None:
+                raise MaxObjectCount(
```
Comment (andy-maier) [inline pywbem/cim_operations.py]: - MaxObjectCount is not an exception.
- In addition, we do not have a CIM status 'CIM_ERR_FILTERED_ENUMERATION_NOT_SUPPORTED'.

The discussion point is whether we want to raise a `CIMError` (which is documented to be an error returned from a server) vs. some other exception Python typically uses for incorrect input parameters, e.g. `ValueError`.

I think I developed an opinion on this: As a user of this function I'd expect a `CIMError(CIM_ERR_NOT_SUPPORTED)` to indicate that the CIMOM or provider did not implement enumeration of instances in any way, so I would find it confusing to raise that when the expectation of the user about pull support was wrong. I'd argue that the user should look at `conn.has_pull_operations` before it attempts to use pull-specific parameters. If they missed that, they deserve that these parameters are considered incorrect input parameters, i.e. I think we should raise a `ValueError` in this case.

File: pywbem/cim_operations.py [hunk 7]
```diff
@@ -1709,6 +1721,269 @@ def OpenEnumerateInstancePaths(self, ClassName, namespace=None,
...
+                # if Not supported, set flag and try alternative operation.
+                except CIMError as ce:
+                    if ce.args[0] != CIM_ERR_NOT_SUPPORTED:
+                        self.has_pull_operations = False
+                    else:
+                        raise
+
+            # Alternate request if Pull not implemented. This does not allow
+            # the FilterQuery
+            if FilterQuery is not None or FilterQueryLanguage is not None:
```
Comment (andy-maier) [inline pywbem/cim_operations.py]: I think `OperationTimeout` should also be rejected.

File: pywbem/cim_operations.py [hunk 8]
```diff
@@ -1709,6 +1721,269 @@ def OpenEnumerateInstancePaths(self, ClassName, namespace=None,
...
+
+                    for inst in pull_result.instances:
+                        yield inst
+
+                    # loop to pull while more exist.
+                    while not pull_result.eos:
+                        pull_result = self.PullInstancesWithPath(
+                            pull_result.context, MaxObjectCount=MaxObjectCount)
+
+                        while inst in pull_result.instances:
```
Comment (andy-maier) [inline pywbem/cim_operations.py]: `while` -> `for`

File: pywbem/cim_operations.py [hunk 9]
```diff
@@ -1709,6 +1721,269 @@ def OpenEnumerateInstancePaths(self, ClassName, namespace=None,
...
+                ClassName, namespace=namespace, LocalOnly=LocalOnly,
+                DeepInheritance=DeepInheritance,
+                IncludeQualifiers=IncludeQualifiers,
+                IncludeClassOrigin=IncludeClassOrigin,
+                PropertyList=PropertyList, **extra)
+                
+            for inst in enum_rslt:
+                yield inst
+
+        # used to clean up if close() executed
```
Comment (andy-maier) [inline pywbem/cim_operations.py]: Don't understand this comment. The code in `finally` gets executed in all cases where one of the operations fails or exhausts the result set. This code is not triggered by the user calling `close()` on the returned iterator.

File: pywbem/cim_operations.py [hunk 10]
```diff
@@ -1709,6 +1721,269 @@ def OpenEnumerateInstancePaths(self, ClassName, namespace=None,
...
+        finally:
+            if pull_result is not None:
+                if not pull_result.eos:
+                    self.CloseEnumeration(pull_result.eos)
```
Comment (andy-maier) [inline pywbem/cim_operations.py]: `pull_result.eos` -> `pull_result.context`

File: pywbem/cim_operations.py [hunk 11]
```diff
@@ -1709,6 +1721,269 @@ def OpenEnumerateInstancePaths(self, ClassName, namespace=None,
                 self.operation_recorder.record_staged()
             return result_tuple
 
+    # TODO overriding the internal choice of pull vs enum. Should we be able
+    #      force one or the other for a connection?
+    # TODO is this a good name for the generator function
+    def EnumerateInstancesGen(self, ClassName, namespace=None,
```
Comment (andy-maier) [inline pywbem/cim_operations.py]: Some material on discussing naming conventions for generator functions:
- https://bytes.com/topic/python/answers/506531-generator-naming-convention
- https://stackoverflow.com/questions/9039265/name-of-a-function-returning-a-generator

I think given that we have different flavors of methods, we cannot just use no prefix or no suffix. It seems to me that using an `iter` prefix would be most Pythonese. Maybe in mixed case to be consistent with our naming style, i.e. `Iter`. Basically, use the name of the open operation and replace `Open` with `Iter`.

File: pywbem/cim_operations.py [hunk 12]
```diff
@@ -1709,6 +1721,269 @@ def OpenEnumerateInstancePaths(self, ClassName, namespace=None,
                 self.operation_recorder.record_staged()
             return result_tuple
 
+    # TODO overriding the internal choice of pull vs enum. Should we be able
+    #      force one or the other for a connection?
```
Comment (andy-maier) [inline pywbem/cim_operations.py]: On overriding: No. The user can already control the behavior of the generator method using the `has_pulled_operations` capability of the connection. If the user wants a different behavior on that same connection than the behavior selected with `has_pulled_operations`, they can use the specific pulled or non-pulled operation methods directly. First, there is no performance disadvantage of doing that compared to using the generator method, because the generator method still materializes the result set of each pull or (in case of non-pulled) of the entire operation, before it yields its instances. Admittedly, a little more coding is needed for using the pull operations directly. Second, cases where it makes sense to use the opposite of what was selected when creating the connection should be rare.

File: pywbem/cim_operations.py [hunk 13]
```diff
@@ -453,11 +453,18 @@ class WBEMConnection(object):
...
+        It can be set by the attribute or if the default of None is set, the
+        first request that uses the pull operations sets the value to True.
+        If set to False, the generator functions will default to using the
+        non-pull equivalent methods
     """
 
     def __init__(self, url, creds=None, default_namespace=DEFAULT_NAMESPACE,
                  x509=None, verify_callback=None, ca_certs=None,
-                 no_verification=False, timeout=None):
+                 no_verification=False, timeout=None, has_pull_operations=None):
```
Comment (andy-maier) [inline pywbem/cim_operations.py]: This parameter does not indicate what the server has, but what this connection wants to use, and we even adjust that as we go, to what the server has.

So I suggest renaming this parameter and the corresponding instance variable to `use_pull_operations`.

**Comment by andy-maier** [comment]
See my comments in the code.

File: pywbem/cim_operations.py (hunk 2 above)
Comment (KSchopmeyer) [inline pywbem/cim_operations.py]: Extra quote

File: pywbem/cim_operations.py (hunk 10 above)
Comment (KSchopmeyer) [inline pywbem/cim_operations.py]: Actually wanted to use the pull_result and use None as a flag.  I did not set it back to None when done however

File: pywbem/cim_operations.py (hunk 1 above)
Comment (KSchopmeyer) [inline pywbem/cim_operations.py]: Used insts_iterator

File: pywbem/cim_operations.py (hunk 5 above)
Comment (KSchopmeyer) [inline pywbem/cim_operations.py]: Will do

Comment (KSchopmeyer) [inline pywbem/cim_operations.py]: Done

File: pywbem/cim_operations.py (hunk 7 above)
Comment (KSchopmeyer) [inline pywbem/cim_operations.py]: Since these do not cause any issues or really change the performance if enumerates are issue I propose that we not flag them.  The goal is to provide as compatible an interface as possible so that the maximum number of parameters do not have to be set dependent on the  type of call, not to detail out each difference.
The timeout simply does not matter if EnumerateInstances is used. It only applies to the pull.

The continue on error does have a user impact. 

I propose test the continueOnError but not the timeout.

File: pywbem/cim_operations.py (hunk 8 above)
Comment (KSchopmeyer) [inline pywbem/cim_operations.py]: Done

File: pywbem/cim_operations.py (hunk 13 above)
Comment (KSchopmeyer) [inline pywbem/cim_operations.py]: Right now I kept them both but want to look at this again

**Comment by andy-maier** [comment]
PR #546 addresses the issue the one Travis CI run encountered by chance. I restarted the run.

File: pywbem/cim_operations.py [hunk 14]
```diff
@@ -1709,6 +1741,299 @@ def OpenEnumerateInstancePaths(self, ClassName, namespace=None,
...
+
+          A Python iterator object. Instances can be retrieved
+          by iterating through the iterator.
+
+        Example:
+            insts_iterator = conn.IterEnumerateInstances('CIM_Blah')
+            for inst in insts_iterator:
+                # close if a particular property value found
+                if inst.get('thisPropertyName') == 0
+                    enum_get.close()
```
Comment (andy-maier) [inline pywbem/cim_operations.py]: Change old variable name `enum_get` to `insts_iterator`.

File: pywbem/cim_operations.py [hunk 15]
```diff
@@ -1709,6 +1741,299 @@ def OpenEnumerateInstancePaths(self, ClassName, namespace=None,
...
+                    break;
```
Comment (andy-maier) [inline pywbem/cim_operations.py]: Semicolon not needed after `break`.

File: pywbem/cim_operations.py [hunk 16]
```diff
@@ -1709,6 +1741,299 @@ def OpenEnumerateInstancePaths(self, ClassName, namespace=None,
...
+                else:
+                    print('instance %s' % inst.tomof())
+
+        """
+
+
+        if MaxObjectCount is not None and MaxObjectCount == 0:
+            raise ValueError('MaxObjectCount=%s. Invalid leq 0' % \
+                             MaxObjectCount)
```
Comment (andy-maier) [inline pywbem/cim_operations.py]: The message still does not reflect the comparison. In addition, we should not use abbreviations such as `leq`.

File: pywbem/cim_operations.py (hunk 3 above)
Comment (andy-maier) [inline pywbem/cim_operations.py]: Done - Was changed as proposed.

File: pywbem/cim_operations.py [hunk 17]
```diff
@@ -453,11 +459,23 @@ class WBEMConnection(object):
...
+        Attribute that defines whether the WBEM server defined by this
+        connection has or is to use the pull operations
+
+        if `None` no pull operation has been attempted on this connection
+        If `True` the user has specified pull operations only or the client
+        has received a successful response to a Iter... method.
+        If `False` the  the user set use_pull_operations=False in the WBEM
+        connection creation or the server has responded with not supported
+        indicating that the pull operations are not supported on the server.
+
```
Comment (andy-maier) [inline pywbem/cim_operations.py]: I can meanwhile understand why you had named the instance variable `has_pull_operations` at some point (if i understood you correctly in the call): After the first call to an `Iter...()` method it is set to reflect the capability of the server, but only if it was not initially passed as `False`. I think naming the instance variable `use_pull_operations` like it is currently is better, because it does not truly reflect the server capability, but what is used in the `Iter...()` methods. No change needed, I just wanted to provide my thoughts on this.

Comments on the description:

1. On "... the WBEM server ... has or is to use ...": It is the client that uses them, not the server.

2. The description of the three values mostly repeats the algorithm that is implemented. I think it should be written by focusing on the use of pull.

3. As an instance variable, I prefer wording that explains what the variable indicates instead of what it controls or defines.

How about this description:

```
        Indicates whether the client will attempt the use of pull operations in any `Iter...()` methods.

        This variable is initially set from the same-named input argument, and is
        adjusted when `Iter...()` methods are invoked, dependent on the support
        for pull operations in the WBEM server.
        After adjusting the variable, it is either `True` or `False` and indicates whether
        pull operations will be used in any `Iter...()` methods.
```

File: pywbem/cim_operations.py [hunk 18]
```diff
@@ -643,6 +661,19 @@ def __init__(self, url, creds=None, default_namespace=DEFAULT_NAMESPACE,
...
+            or only the non-pull operations for the Iter... methods. Normally
+            this should NOT be used but it may prove useful in some test or
+            error cases.
+            If `None` ignored and the server sets the attribute based on the
+            response to the first pull operation.
+            If `True` forces the client to use only the pull operations for
+            the Iter... methods and return any response  to a pull operation
+            including `CIM_ERR_NOT_SUPPORTED` as an error on this connection.
+            If `False` forces the client to use only the non-pull operations
+            (ex. EnumerateInstances) for the Iter... methods.
```
Comment (andy-maier) [inline pywbem/cim_operations.py]: Comments on the description of the `use_pull_operation`  ctor argument:

1. The "force" only works in one direction, at least in the current implementation: `False` forces non-use even if pull is supported by the server, but `True` specifies that an attempt to use pull is made, which may fall back to using non-pull.

2. On "Normally this should not be used...": It is not clear what the "this" refers to.

3. For `None`, I would like to see a more explicit statement that pull is attempted.

4. For `True`, it does not force the use of only pull. That may have been an a prior implementation, but the current code attempts pull and if it fails with not supported, goes on to attempt non-pull. From a perspective of the (unadjusted) input argument, `None` specifies exactly the same as `True`. Which is fine IMO, just the description needs to be fixed.

5. On "If `False` forces ...": I don't like this super-short form. Better use "`False` forces ...".

How about this description:

```
            Controls the use of pull operations in any `Iter...()` methods. The default
            value will work fine on any WBEM server whether or not it supports pull
            operations.
            `None` and `True` mean that pull operations will be attempted, and if
            the WBEM server does not support pull operations, the corresponding
            instance variable is set to `False`, causing the traditional operations
            to be used in the future.
            `False` means that only the traditional operations will be used,
            regardless of whether or not the WBEM server also supports
            pull operations.
```

File: pywbem/cim_operations.py (hunk 6 above)
Comment (andy-maier) [inline pywbem/cim_operations.py]: Done, ValueError is now raised.

File: pywbem/cim_operations.py (hunk 7 above)
Comment (andy-maier) [inline pywbem/cim_operations.py]: Done: Karl added checking for ContinueOnError, as proposed.

File: pywbem/cim_operations.py [hunk 19]
```diff
@@ -1709,6 +1741,299 @@ def OpenEnumerateInstancePaths(self, ClassName, namespace=None,
...
+                ClassName, namespace=namespace, LocalOnly=LocalOnly,
+                DeepInheritance=DeepInheritance,
+                IncludeQualifiers=IncludeQualifiers,
+                IncludeClassOrigin=IncludeClassOrigin,
+                PropertyList=PropertyList, **extra)
+
+            for inst in enum_rslt:
+                yield inst
+
+        # Clean up if close() executed and  pull operation is active
```
Comment (andy-maier) [inline pywbem/cim_operations.py]: I finally understand the situation this comment describes: The user is in the loop iterating over the items yielded by this method (i.e. iterates over the returned iterator), and for some reason calls close() on the returned iterator before the iterator is exhausted. Using the wording "and pull operation is active" is a bit confusing to me, because it is not the pull operation against the server that is active, but the iterator is not yet exhausted.

Can we change this comment to:

```
# Clean up if the caller closes the iterator before exhausting it
```

File: pywbem/cim_operations.py [hunk 20]
```diff
@@ -1709,6 +1741,299 @@ def OpenEnumerateInstancePaths(self, ClassName, namespace=None,
...
+        finally:
+            if pull_result is not None and not pull_result.eos:
+                self.CloseEnumeration(pull_result.eos)
```
Comment (andy-maier) [inline pywbem/cim_operations.py]: Error: `CloseEnumeration()` must be called with the context.

To fix this, change line 2034 to:

```
self.CloseEnumeration(pull_result.context)
```

File: pywbem/cim_operations.py [hunk 21]
```diff
@@ -1944,6 +2269,9 @@ class are to be included in the returned instances, as follows:
                 MaxObjectCount=MaxObjectCount,
                 **extra)
 
+        if MaxObjectCount is not None and MaxObjectCount < 0:
+            raise ValueError('MaxObjectCount=%s. Invalid lt 0' % MaxObjectCount)
```
Comment (andy-maier) [inline pywbem/cim_operations.py]: Suggest to change the message to be a human readable statement, e.g.:

```
"MaxObjectCount must be >= 0, but is %s" % MaxObjectCount
```

File: pywbem/cim_operations.py (hunk 10 above)
Comment (andy-maier) [inline pywbem/cim_operations.py]: The comment applies to line 1984. I raised this issue again on the latest version.

File: pywbem/cim_operations.py (hunk 11 above)
Comment (andy-maier) [inline pywbem/cim_operations.py]: Done, Karl changed it to `Iter...()`

File: pywbem/cim_operations.py (hunk 12 above)
Comment (andy-maier) [inline pywbem/cim_operations.py]: Done - comment was removed, and no override control is provided.

File: pywbem/cim_operations.py (hunk 13 above)
Comment (andy-maier) [inline pywbem/cim_operations.py]: Done - renamed to `use_pull_operations`.

**Comment by andy-maier** [comment]
I have some more comments on the last commit, see above or the "Files changed" tab.

File: pywbem/cim_operations.py (hunk 14 above)
Comment (KSchopmeyer) [inline pywbem/cim_operations.py]: Done

File: pywbem/cim_operations.py (hunk 16 above)
Comment (KSchopmeyer) [inline pywbem/cim_operations.py]: Done

File: pywbem/cim_operations.py (hunk 17 above)
Comment (KSchopmeyer) [inline pywbem/cim_operations.py]: It was originally just internal, there has ... When we extended it to cover both what the client sees and what the user wants, we sort of mixed things up.

Done

File: pywbem/cim_operations.py (hunk 19 above)
Comment (KSchopmeyer) [inline pywbem/cim_operations.py]: Done. Added comment explaining why only on pull below the finally also

File: pywbem/cim_operations.py (hunk 20 above)
Comment (KSchopmeyer) [inline pywbem/cim_operations.py]: Good catch.

File: pywbem/cim_operations.py (hunk 21 above)
Comment (KSchopmeyer) [inline pywbem/cim_operations.py]: Done.  should be > 0, not >= 0

Comment (KSchopmeyer) [inline pywbem/cim_operations.py]: There were multiples and fixed them all.  Note that since we have to add this test to all open/pulls I will make it a common function but right now concentrating on the Iter...

File: pywbem/cim_operations.py (hunk 15 above)
Comment (andy-maier) [inline pywbem/cim_operations.py]: Done by Karl.

File: pywbem/cim_operations.py [hunk 22]
```diff
@@ -1709,6 +1741,299 @@ def OpenEnumerateInstancePaths(self, ClassName, namespace=None,
...
+            execution of this generator.
+
+            * If positive, the WBEM server is to return no more than the
+              specified number of instances.
+            * Zero is not allowed in that it would mean that zero instances
+              are to be returned for every request issued.
+            * If `None`, this parameter is not passed to the WBEM server, and
+              causes the server-implemented default behaviour to be used.
+              :term:`DSP0200` defines that the server-implemented default is
+              to return zero instances.
```
Comment (andy-maier) [inline pywbem/cim_operations.py]: This one is still open, we discussed in the call today that we would address it as proposed.

File: pywbem/cim_operations.py [hunk 23]
```diff
@@ -453,11 +459,33 @@ class WBEMConnection(object):
...
+        Indicates whether the client will attempt the use of pull operations in
+        any `Iter...()` methods.
+
+        This variable is initially set from the same-named input argument, and
+        is modified when `Iter...()` methods are invoked, dependent on the
+        support for pull operations in the WBEM server.
+        After modifying the variable, it is either `True` or `False` and
+        indicates whether pull operations will be used in any subsequent
+        `Iter...()` methods.
+
```
Comment (andy-maier) [inline pywbem/cim_operations.py]: The new text for the `use_pull_operations` instance variable (above this comment) makes the part of the old text that is till below this comment (lines 474 to 482) superfluous. They should be removed.

**Comment by andy-maier** [comment]
I had three more comments, see the "Files changed" tab.

File: pywbem/cim_operations.py [hunk 24]
```diff
@@ -643,6 +671,24 @@ def __init__(self, url, creds=None, default_namespace=DEFAULT_NAMESPACE,
...
+            not it supports pull operations so that first the pull operation is
+            attempted and if the WBEM server does not support pull operations,
+            the corresponding instance variable is set to `False` the
+            traditional (non-pull) operation will be executed for the current
+            operation and all subsequent  Iter...()operations for this
+            WBEM Connection.
+
+            `True` means that pull operations will be attempted; all errors
+            including `CIM_ERR_NOT_SUPPORTED` are returned to the caller. The
+            traditional operations will not be attempted.
```
Comment (KSchopmeyer) [inline pywbem/cim_operations.py]: No.  The way the code works is that if True is set on the attribute, a failure of the open or pull goes directly to raise in all cases.  The assumption we are making is that if true is set either by the attribute or initial success, all operations MUST be pull.   Unless I misread my code that is what it does.  The only time it first tries pull and then falls back to enum is if use... is None and the error is CIM_ERR_NOT_SUPPORTED. 

                except CIMError as ce:
                    if self.use_pull_operations is None and \
                       ce.status_code == CIM_ERR_NOT_SUPPORTED:
                        self.use_pull_operations = False
                    else:
                        raise

Any other exception gets raised directly.

File: pywbem/cim_operations.py (hunk 23 above)
Comment (KSchopmeyer) [inline pywbem/cim_operations.py]: I agree.  Removed

File: pywbem/cim_operations.py (hunk 24 above)
Comment (andy-maier) [inline pywbem/cim_operations.py]: Karl, you did not miss anything. My reading of the code was wrong. So in that case, I have a few other comments on the description, which I will make in the file again.

File: pywbem/cim_operations.py [hunk 25]
```diff
@@ -453,11 +460,22 @@ class WBEMConnection(object):
...
+      use_pull_operations (:class:`py:bool`):
+        Indicates whether the client will attempt the use of pull operations in
+        any `Iter...()` methods.
+
+        This variable is initially set from the same-named input argument, and
+        is modified when `Iter...()` methods are invoked, dependent on the
+        support for pull operations in the WBEM server.
+        After modifying the variable, it is either `True` or `False` and
+        indicates whether pull operations will be used in any subsequent
+        `Iter...()` methods.
```
Comment (andy-maier) [inline pywbem/cim_operations.py]: In order to reflect the fact that only for an initial value of `None`, the instance variable will be adjusted, I suggest to replace the paragraph above with this:

```
        This variable is initially set from the same-named input argument. If
        it is `None`, it is adjusted when `Iter...()` methods are invoked, dependent
        on the support for pull operations in the WBEM server. After the first call
        to an `Iter...()` method, the variable will be either `True` or `False`.
```

File: pywbem/cim_operations.py [hunk 26]
```diff
@@ -643,6 +661,24 @@ def __init__(self, url, creds=None, default_namespace=DEFAULT_NAMESPACE,
...
+
+          use_pull_operations (:class:`py:bool`):
+            Controls the use of pull operations in any `Iter...()` methods.
+            The default value (`None`) will work on any WBEM server whether or
+            not it supports pull operations; first the pull operation is
+            attempted and if the WBEM server does not support pull operations,
+            the corresponding instance variable is set to `False` and the
+            traditional (non-pull) operation will be executed for the current
+            operation and all subsequent Iter...()operations for this
+            WBEM Connection.
```
Comment (andy-maier) [inline pywbem/cim_operations.py]: The paragraph above is not explicit about the fact that the instance variable is set to True is the server supports pull operations. I suggest to replace it with this text:

```
            Controls the use of pull operations in any `Iter...()` methods.
            The default value (`None`) will work on any WBEM server whether or
            not it supports pull operations.

            `None` means that first the pull operation is attempted. If that
            succeeds, corresponding instance variable is set to `True`. If the
            WBEM server does not support pull operations, the corresponding
            instance variable is set to `False` and the traditional (non-pull)
            operation will be executed.
```

File: pywbem/cim_operations.py [hunk 27]
```diff
@@ -643,6 +661,24 @@ def __init__(self, url, creds=None, default_namespace=DEFAULT_NAMESPACE,
...
+
+            `True` means that pull operations will be attempted; all errors
+            including `CIM_ERR_NOT_SUPPORTED` are returned to the caller. The
+            traditional operations will not be attempted.
+
+            `False` means that only the traditional operations will be used,
+            regardless of whether or not the WBEM server also supports
+            pull operations.
```
Comment (andy-maier) [inline pywbem/cim_operations.py]: The description is not explicit about the fact that the instance variable is set to True is the server supports pull operations, and we could do a few editorial improvements as well.

I suggest to replace the description with this one:

```
            Controls the use of pull operations in any `Iter...()` methods, by
            intializing a corresponding instance variable that is used during
            any `Iter...()` methods.
            The default value (`None`) will work on any WBEM server whether or
            not it supports pull operations.

            `None` means that the `Iter...()` methods will attempt a pull operation
            first, and if the WBEM server does not support it, will use a traditional
            operation. The corresponding instance variable will be adjusted to
            indicate what the WBEM server supports.

            `True` means that the `Iter...()` methods will only use pull operations.
            If the WBEM server does not support pull operations, a
            :exc:`~pywbem.CIMError` with status code `CIM_ERR_NOT_SUPPORTED`
            will be raised.

            `False` means that the `Iter...()` methods will only use traditional operations.
```

File: pywbem/cim_operations.py [hunk 28]
```diff
@@ -1709,6 +1745,311 @@ def OpenEnumerateInstancePaths(self, ClassName, namespace=None,
...
+
+        This method performs either the `OpenEnumerationInstances` and
+        `PullInstancesWithPath` operations (pull requests) or the
+        EnumerateInstances operation if the WBEM server does not support the
+        DMTF pull operations. It is an alternate to using `EnumerateInstances`
+        or a `OpenEnumerateInstances`/`PullInstancesWithPath` sequence.
+
+        The `use_pull_operations` attribute of the WBEMConnection constructor
+        may be used to control whether pull or traditional requests are
+        used.
```
Comment (andy-maier) [inline pywbem/cim_operations.py]: The two paragraphs above could be improved to talk about the default behavior and that the input argument can be used to override the default. Also, I would use "argument" and not "attribute".
Finally, links could be used where possible.

To address all of that, I suggest to replace the two paragraphs above with these:

```
        This method performs either the
        :meth:`~pywbem.WBEMConnection.OpenEnumerationInstances` and
        :meth:`~pywbem.WBEMConnection.PullInstancesWithPath`
        operations (pull operations) or the
        :meth:`~pywbem.WBEMConnection.EnumerateInstances` operation
        (traditional operation) if the WBEM server does not support the
        pull operations. It is an alternative to using these operations that
        automatically uses the pull operations if supported.

        The `use_pull_operations` argument of the
        :class:`~pywbem.WBEMConnection` constructor can be used to
        override that default behavior to either insist on pull operations or
        to avoid their use.
```

**Comment by andy-maier** [comment]
Some more comments in the "Files changed" section.

Plus this one:

* By having only one variable for the use of pull operations, we assume that a WBEM server has either implemented all pull operations or none of them. An implementation of only OpenEnumerateInstances() but not of "OpenAssociators()" and "OpenReferences()" will cause failures, because we are setting the instance variable to True if the first pull operation in any Iter...() method succeeds.

  This problem becomes more likely considering that CIM_ERR_NOT_SUPPORTED is probably also returned if the provider does not support the attempted pull operation. That could even happen when the CIMOM transforms the pull client operation into a non-pull provider operation, and the provider simply does not support it (e.g. an association provider supports Associators but not References).

**Discussion point:** Does the client need to accomodate for the case that a WBEM server supports only a subset of the pull operations?

**Comment by KSchopmeyer** [comment]
One more interesting comment. We have a public attribute use_pull_operations.  The user can reset that at any time to start over.  In fact I am using it in the tests to first force pull and then non-pull rather than just setting it when the connection is made. That is just a side effect but it is the api right now

**Comment by KSchopmeyer** [comment]
We propose to just extend the documentation so that users understand the potential issue.

**Comment by andy-maier** [comment]
To get the Travis issues fixed, please rebase this on branch `issue550-pin-gitpython-208`.

**Comment by andy-maier** [comment]
I have merged PR #551 and deleted branch `issue550-pin-gitpython-208`, so you now need to rebase on `master`.

**Comment by KSchopmeyer** [comment]
Discussion:  The individual flags.

**Comment by andy-maier** [comment]
Result of the discussion about the individual flags:
- code sets the individual flags from the connection level flag. The individual flags get updated based upon what the server supports. The individual flags are not fed back to update the connection level flag.
- there is a new function set_enumerate...() that allows changing the connection level flag and essentially starts over again with the automatic updating of the individual flags.

**Comment by andy-maier** [comment]
Remaining Todos:
* Review whether we want the set_enumerate....() function
* Review entire code
Once both of that has happened:
* Implement the rest of the functions.

**Comment by KSchopmeyer** [comment]
Updated to rebase to master that has all flake8 fixes and rebase into single set of changes.

**Comment by andy-maier** [comment]
On the need for the `set_use_pull_operations()` method: The CIMOM's support for pull operations does not change for a given connection. However, provider support may vary across the different providers.

If a CIMOM compensates for that, i.e. if it services a client's pull operation by invoking the classic provider operation in case a provider does not support pull, then we don't need the method. Unfortunately, the need for such a compensating behavior is not specified anywhere.

Still, I have a hard time imagining a client program using the `set_use_pull_operations()` method to adjust to pull vs. classic at a provider granularity, and possibly different for each server. Having said that, the alternative to using the method is to have one connection with, and one without pull enabled, and to use them dependent on the provider implementation. That is not any less ugly, but also not significantly more ugly, IMO. The ugliness is not so much the second connection, but the need to understand which provider on which type of server environment implemented pull.

In other words, the `set_use_pull_operations()` method does not reduce the ugliness in any significant way.

I believe the way out is the deprecation of the classic operations in DSP0200. That will cause CIMOM implementers to support the compensating behavior described above.

Overall, I think we should not have the `set_use_pull_operations()` method.

**Comment by andy-maier** [comment]
Otherwise, +1 on the implemented operations.

**Comment by KSchopmeyer** [comment]
We agreed to add the missing path information back in so old and new operations match

Also proposed as separate issue to fix the original operations based on a config variable to that paths include namespace and host.

**Comment by KSchopmeyer** [comment]
We think this should be a higher level class, ie. IterWBEMConnection.  We will do that before commit.

We also concluded that the recorder is at the lowest level.

File: pywbem/cim_operations.py [hunk 29]
```diff
@@ -482,11 +534,26 @@ class WBEMConnection(object):
         :class:`~pywbem.BaseOperationRecorder`, each operation
         that is executed on this connection will be recorded by
         invoking its :meth:`~pywbem.BaseOperationRecorder.record` method.
```
Comment (andy-maier) [inline pywbem/cim_operations.py]: This is where we can add a statement that for the `Iter*()` methods, the recording will be done at the level of the lower level methods they invoke.

**Comment by andy-maier** [comment]
I agree with the approach, and have made a comment where I suggested a good place for adding the statement about recording for the Iter functions.

**Comment by andy-maier** [comment]
+1
